### PR TITLE
ROC-2300: Add sonar properties

### PR DESCRIPTION
**Pull request by vianvi**
Resolves # . (This is applicable only if this pull request relates to a GitHub issue, delete the line otherwise)

Notes:
*
*
*

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -95,15 +96,16 @@
     "sinon": "^3.2.1",
     "sinon-chai": "^2.13.0",
     "sinon-express-mock": "^1.3.0",
+    "sonar-scanner": "^3.0.3",
     "supertest": "^3.0.0",
     "tslint": "^5.5.0",
     "tslint-config-standard": "^6.0.1"
   },
   "nyc": {
     "include": [
```
Comment (dharmendrak) [inline package.json]: why did we had test before? why it's no more used

Comment (vianvi) [reply package.json]: if you mean in  nyc area - yes it was not configured in my earlier pull request

File: package.json [hunk 2]
```diff
@@ -17,7 +17,8 @@
     "test:routes": "NODE_ENV=mocha LOG_LEVEL=OFF mocha --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) -path '*/routes/*') --reporter mochawesome --reporter-options reportFilename=routes,inlineAssets=true,reportTitle=legal-rep-frontend-routes",
     "test:a11y": "mocha --compilers ts:ts-node/register -r tsconfig-paths/register src/test/a11y --timeout 60000 --reporter mochawesome --reporter-options reportFilename=a11y,inlineAssets=true,reportTitle=legal-rep-frontend-a11y",
     "test:nsp": "nsp check",
-    "test:coverage": "cross-env NODE_PATH=. nyc istanbul cover  ./node_modules/mocha/bin/_mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*' ! -path '*/routes/*')",
+    "test:coverage": "cross-env NODE_PATH=. nyc istanbul cover  ./node_modules/mocha/bin/_mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*')",
+    "sonar-scanner": "cross-env NODE_PATH=. node_modules/sonar-scanner/bin/sonar-scanner",
```
Comment (timja-kainos) [inline package.json]:
```
cross-env NODE_PATH=. sonar-scanner
```

File: package.json [hunk 3]
```diff
@@ -17,7 +17,8 @@
     "test:routes": "NODE_ENV=mocha LOG_LEVEL=OFF mocha --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) -path '*/routes/*') --reporter mochawesome --reporter-options reportFilename=routes,inlineAssets=true,reportTitle=legal-rep-frontend-routes",
     "test:a11y": "mocha --compilers ts:ts-node/register -r tsconfig-paths/register src/test/a11y --timeout 60000 --reporter mochawesome --reporter-options reportFilename=a11y,inlineAssets=true,reportTitle=legal-rep-frontend-a11y",
     "test:nsp": "nsp check",
-    "test:coverage": "cross-env NODE_PATH=. nyc istanbul cover  ./node_modules/mocha/bin/_mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*' ! -path '*/routes/*')",
+    "test:coverage": "cross-env NODE_PATH=. nyc istanbul cover  ./node_modules/mocha/bin/_mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*')",
```
Comment (timja-kainos) [inline package.json]:
```
cross-env NODE_PATH=. nyc istanbul cover  mocha ...
```

File: package.json [hunk 4]
```diff
@@ -123,7 +127,7 @@
       "text",
       "text-summary"
     ],
-    "all": true,
-    "check-coverage": true
+    "report-dir": "test/coverage",
```
Comment (sabahirfan) [inline package.json]: could we please move this under coverage dir rather than on `test/coverage`?

**Comment by sabahirfan** [comment]
Please update README and add the coverage command if we want to run it locally.

File: package.json (hunk 4 above)
Comment (vianvi) [reply package.json]: Sure will do

File: package.json (hunk 3 above)
Comment (timja-kainos) [reply package.json]: what's the error?

Comment (vianvi) [reply package.json]: Something like this: so i have used full path for running from windows machine for mocha and when i finished my local test updated back to as shown above

C:\cmc\cmc-legal-rep-frontend\node_modules\.bin\mocha.CMD:1
(function (exports, require, module, __filename, __dirname) { @IF EXIST "%~dp0\node.exe" (
                                                              ^

SyntaxError: Invalid or unexpected token
    at createScript (vm.js:74:10)
    at runInThisContext (vm.js:116:10)
    at Object.vm.runInThisContext (C:\CMC\cmc-legal-rep-frontend\node_modules\nyc\node_modules\istanbul-lib-hook\lib\hook.js:141:16)
    at Module._compile (module.js:537:28)
    at Module.replacementCompile (C:\CMC\cmc-legal-rep-frontend\node_modules\nyc\node_modules\append-transform\index.js:58:13)
    at module.exports (C:\CMC\cmc-legal-rep-frontend\node_modules\nyc\node_modules\default-require-extensions\js.js:8:9)
    at Object..js (C:\CMC\cmc-legal-rep-frontend\node_modules\nyc\node_modules\append-transform\index.js:62:4)
    at Module._extensions.(anonymous function) (C:\CMC\cmc-legal-rep-frontend\node_modules\istanbul\lib\hook.js:109:37)
    at Object.<anonymous> (C:\CMC\cmc-legal-rep-frontend\node_modules\nyc\node_modules\append-transform\index.js:62:4)
    at Module.load (module.js:507:32)

File: package.json [hunk 5]
```diff
@@ -17,7 +17,8 @@
     "test:routes": "NODE_ENV=mocha LOG_LEVEL=OFF mocha --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) -path '*/routes/*') --reporter mochawesome --reporter-options reportFilename=routes,inlineAssets=true,reportTitle=legal-rep-frontend-routes",
     "test:a11y": "mocha --compilers ts:ts-node/register -r tsconfig-paths/register src/test/a11y --timeout 60000 --reporter mochawesome --reporter-options reportFilename=a11y,inlineAssets=true,reportTitle=legal-rep-frontend-a11y",
     "test:nsp": "nsp check",
-    "test:coverage": "cross-env NODE_PATH=. nyc istanbul cover  ./node_modules/mocha/bin/_mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*' ! -path '*/routes/*')",
+    "test:coverage": "cross-env NODE_PATH=. NODE_ENV=mocha nyc istanbul cover mocha -- --compilers ts:ts-node/register -r tsconfig-paths/register $(find src/test \\( -name '*.ts' -or -name '*.js' \\) ! -path '*a11y*')",
```
Comment (timja-kainos) [inline package.json]:
```
yarn check
```
fails when run on nyc dependency.

It is used to make sure the lock file and package.json file is in sync

T2 is missing the check that enforces it in their CI,
example from T1
https://github.com/hmcts/cmc-citizen-frontend/blob/master/.travis.yml#L6

File: package.json (hunk 3 above)
Comment (timja-kainos) [reply package.json]: This is the istanbul issue:
https://github.com/gotwarlost/istanbul/issues/13#issuecomment-18175599

File: package.json (hunk 5 above)
Comment (timja-kainos) [reply package.json]: @vianvi This is still failing and if I try and my console gets hammered with warnings like this:
```
warning The case-insensitive file /Users/timja/projects/moj-reform/cmc/cmc-legal-rep-frontend/node_modules/nyc/node_modules/yargs/node_modules/yargs-parser/package.json shouldn't be copied twice in one bulk copy
```

File: README.md [hunk 6]
```diff
@@ -60,3 +60,18 @@ For accessibilit testing:
...
 $ yarn test:a11y
 ```
+For code coverage:
+```bash
+$ yarn test:coverage
+$ yarn sonar-scanner
+```
+## Troubleshooting
+
+### nyc warings while running  ```yarn install``` on yarn version 1.0.1
```
Comment (timja-kainos) [inline README.md]: ### Warnings while running  ```yarn install``` on yarn version 1.0.1

File: README.md [hunk 7]
```diff
@@ -60,3 +60,18 @@ For accessibilit testing:
...
+
+This is currently an Open issue : https://github.com/yarnpkg/yarn/issues/3751
```
Comment (timja-kainos) [inline README.md]: There is currently an open issue,
also include a short example of the warnings only one or two lines

File: README.md [hunk 8]
```diff
@@ -60,3 +60,18 @@ For accessibilit testing:
...
+and also for checking integrity use following command instead of ```yarn check```
```
Comment (timja-kainos) [inline README.md]: this section isn't needed

File: README.md [hunk 9]
```diff
@@ -60,3 +60,20 @@ For accessibilit testing:
 ```bash
 $ yarn test:a11y
 ```
+For code coverage:
+```bash
+$ yarn test:coverage
+$ yarn sonar-scanner
```
Comment (timja-kainos) [inline README.md]: I don't think people should be running this locally?
yarn test:coverage should be enough?

Comment (timja-kainos) [reply README.md]: Also missing changes adding this into Jenkins

Comment (vianvi) [reply README.md]: We have two separate tickets for addressing above problems
ROC-2325
ROC-2326

Comment (timja-kainos) [reply README.md]: regardless `yarn sonar-scanner` shouldn't be used by developers so it shouldn't be here

Comment (vianvi) [reply README.md]: as part of 2326 we can add this to stage to jenkins

Comment (timja-kainos) [reply README.md]: you're adding a line to the README that shouldn't be there, that's not how developers should be seeing code coverage.

`$ yarn test:coverage` is

Comment (vianvi) [reply README.md]: can you suggest me how you want me to change README?

File: sonar-project.properties [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+sonar.sources=src/main
+sonar.tests=src/test
+sonar.language=ts
+sonar.exclusions=src/test/**, src/main/resources/**, src/main/public/**/*,"src/main/views/*.njk", src/main/views/**, src/main/resources/**, src/main/modules/**
+sonar.dynamicAnalysis=reuseReports
+
+sonar.sourceEncoding=UTF-8
+sonar.ts.tslint.configPath=tslint.json
+sonar.ts.coverage.lcovReportPath=coverage/lcov.info
+sonar.host.url=https://sonar.reform.hmcts.net
```
Comment (timja-kainos) [inline sonar-project.properties]: I don't think this should be set here

Comment (vianvi) [reply sonar-project.properties]: I am not sure how we can set, just I followed footsteps of Probate and SSCS Frontend projects

Comment (timja-kainos) [reply sonar-project.properties]: probably can do `sonar-scanner -Dsonar.host.url=${ENV_URL}`
Something like this:
https://github.com/hmcts/cmc-claim-store/blob/master/Jenkinsfile#L85

Comment (geek-soft) [reply sonar-project.properties]: Good spot @timja-kainos. It is possible by providing `SONAR_SCANNER_OPTS` e.g.

```
export SONAR_SCANNER_OPTS="-Dsonar.host.url=$SONARQUBE_URL"
```

File: sonar-project.properties [hunk 11]
```diff
@@ -0,0 +1,14 @@
+sonar.projectKey=CMCLFE
+sonar.projectName=CMC :: Legal Frontend
+sonar.projectVersion=0.0.1
```
Comment (geek-soft) [inline sonar-project.properties]: Did ask to remove this as well as this makes maintenance just harder. On Jenkins to could be probably overridden.

File: sonar-project.properties (hunk 10 above)
Comment (vianvi) [reply sonar-project.properties]: removed the host url and when we do ticket ROC-2326 we add this to jenkins stage

File: sonar-project.properties (hunk 11 above)
Comment (vianvi) [reply sonar-project.properties]: removed version as well

**Pull request merged**